### PR TITLE
Exclude URLs feature

### PR DESCRIPTION
**Pull request by stephskuola**
implemented exclude Urls feature: now you can declare into the sitemap.yml a regex pattern to exclude some URls from the sitemap

### CODE REVIEW

File: Command/SitemapGeneratorCommand.php [hunk 1]
```diff
@@ -56,13 +56,22 @@ class SitemapGeneratorCommand extends ContainerAwareCommand
...
+
+    /**
      * SitemapGeneratorCommand constructor.
      *
      * @param RouterInterface $router
      * @param ObjectManager   $objectManager
      * @param array           $config
      */
-    public function __construct(RouterInterface $router, ObjectManager $objectManager, array $config)
+    public function __construct(
```
Comment (fabiocicerchia) [inline Command/SitemapGeneratorCommand.php]: It is better to revert this formatting change and just keep the changes minimal.

File: Command/SitemapGeneratorCommand.php [hunk 2]
```diff
@@ -148,6 +157,14 @@ protected function runMultiSitemaps()
      */
     protected function generateSitemap($name, array $options)
     {
+        if(
+            isset($options['exclude_urls']) &&
```
Comment (fabiocicerchia) [inline Command/SitemapGeneratorCommand.php]: There's no need to use `isset` in combination with `empty`

File: Command/SitemapGeneratorCommand.php [hunk 3]
```diff
@@ -200,12 +217,21 @@ protected function generateSitemapFromRoutes(Sitemap $sitemap, array $routes, $m
             );
 
             if (empty($routeParameters)) {
-                $sitemap->addItem(
-                    $this->router->generate($routeName, [], true),
-                    $routeConfigurations['lastmod'],
-                    $routeConfigurations['changefreq'],
-                    $routeConfigurations['priority']
-                );
+                $curRouteUrl = $this->router->generate($routeName, [], true);
```
Comment (fabiocicerchia) [inline Command/SitemapGeneratorCommand.php]: I'd suggest changing the name of this variable to `$currentRouteUrl`

File: Command/SitemapGeneratorCommand.php [hunk 4]
```diff
@@ -200,12 +217,21 @@ protected function generateSitemapFromRoutes(Sitemap $sitemap, array $routes, $m
...
+
+                if (!$this->isUrlExcluded($curRouteUrl)) {
```
Comment (fabiocicerchia) [inline Command/SitemapGeneratorCommand.php]: You can also remove the else statement by flipping the `if` condition, like this:

```php
if ($this->isUrlExcluded($curRouteUrl)) {
    $this->output->writeln('this URL is excluded by exclude_urls param: ' . $curRouteUrl);
}
```

File: Command/SitemapGeneratorCommand.php [hunk 5]
```diff
@@ -262,20 +288,30 @@ protected function addItems(Sitemap $sitemap, $route, array $parametersCollectio
...
         foreach ($parametersCollection as $parameters) {
-            $sitemap->addItem(
-                $this->router->generate($route, $this->accessor->getValue($parameters, '[route_params]') ?: $parameters, true),
-                $this->accessor->getValue($parameters, '[sitemap_optional_tags][lastmod]')    ?: $routeConfigurations['lastmod'],
-                $this->accessor->getValue($parameters, '[sitemap_optional_tags][changefreq]') ?: $routeConfigurations['changefreq'],
-                $this->accessor->getValue($parameters, '[sitemap_optional_tags][priority]')   ?: $routeConfigurations['priority']
-            );
+            $curRouteUrl = $this->router->generate($route, $this->accessor->getValue($parameters, '[route_params]') ?: $parameters, true);
+
+            if (!$this->isUrlExcluded($curRouteUrl)) {
```
Comment (fabiocicerchia) [inline Command/SitemapGeneratorCommand.php]: You can flip the `if` condition here as well

File: DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -50,6 +50,9 @@ public function getConfigTreeBuilder()
                             ->scalarNode('path')
                                 ->defaultValue('%kernel.root_dir%/../web/sitemap.xml')
                             ->end()
+                            ->scalarNode('exclude_urls')
+                                ->defaultValue('')
```
Comment (fabiocicerchia) [inline DependencyInjection/Configuration.php]: As per `protected $excludeUrlsPattern = null;` the default value should be `null`

File: Command/SitemapGeneratorCommand.php [hunk 7]
```diff
@@ -461,4 +497,28 @@ protected function getRouteParamaters(array $routeConfigurations)
...
         );
     }
+
+    /**
+     * @param string $url
+     * @return bool
+     */
+    protected function isUrlExcluded($url)
+    {
+        $isExcluded = false;
```
Comment (fabiocicerchia) [inline Command/SitemapGeneratorCommand.php]: Would be good to remove as well this extra variable.

Example:
```php
if (empty($this->excludeUrlsPattern)) {
    return false;
}
```

and

```php
$result = preg_match($this->excludeUrlsPattern, $url);
return (bool) $result;
```

**Review by fabiocicerchia — CHANGES_REQUESTED** [review]

**Comment by massimo-me** [comment]
Why you don't uses a custom provider for exclude your business logic?

If I have 200k urls and only one excluded url. I'll have 200k `preg_match` executions 🤔

**Comment by fabiocicerchia** [comment]
@ChiarilloMassimo That's a good point, how would you suggest to do it? By filtering the URL list upfront, maybe right in the SQL query?

**Comment by Harumaro** [comment]
What about

```php
$patterns = '/'.str_replace('/', '', implode('|', $this->excludeUrlPatterns)).'/';
$result = preg_match($patterns, $url);
```

Should solve the scalability issue.

**Comment by massimo-me** [comment]
@fabiocicerchia 

e.g

```php
use Skuola\SitemapBundle\Service\ParametersCollectionInterface;

class PageProvider implements ParametersCollectionInterface {
    protected $entityManager;

    public function __construct($entityManager)
    {
        $this->entityManager = $entityManager;
    }

    public function getParametersCollection() {
        $collection = [];
        
        //Custom find
        $pages = $this->entityManager->getRepository('Page')->find.....();

        foreach($pages as $page) {

            // Custom logic
            // if (preg_match('/^admin\/*/', $page->getPrettyUrl()); { continue }

            $collection[] = [
               'category_slug' => $page->getCategory()->getSlug(),
               'page_slug'     => $page->getSlug()
            ]
        }

        return $collection;
    }
}
```

**Pull request closed without merging**